### PR TITLE
Refactoring: Remove 'prodtype' Mk.2

### PR DESCRIPTION
**Pull request by evgenyz**
This is the replacement for #10726. Check https://github.com/ComplianceAsCode/content/issues/11382 for the context of this change.

#### Description:

This is the first step to get rid of prodtype, it introduces: 

1) A way to ignore current value of prodtype during the build phase
2) A mechanism that would filer out rules not mentioned in a profile (control file)
3) A profile property that allows to 'hide' it in the XCCDF Benchmark
4) A prodtype -> profile entry converter `utils/convert_prodtype_into_default_profile_entry.py`
5) A prodtype cleaner `utils/convert_prodtype_remove_prodtype.sh`
6) A script that analyses a DS from rule/profile/group/selection POV `utils/convert_prodtype_ds_check.py`
7) The actual changes to `rule.yml/group.yml` and `profiles/*.profile` files

#### Rationale:

To confine the amount of changes for testing and reviewing we first implement the new behaviour mostly ignoring the existence of prodtype resolving. After this PR will be merged new way of rule linking will be in effect immediately. The removal of prodtype pieces will be implemented in a different PR.

This PR is not yet complete because it has to have the conversion: removal of prodtypes from rules and rules records in profiles. I left it out to simplify reviewing process. Once we agree that the approach is sound (and/or fix something) I'll rebase the branch to catch all the updated rules, run the converter and add these changes back to this PR to be merged.

This PR should not disrupt anything in the content from customer and SME point of view: all rules, all profiles and rules selected/deselected state should stay exactly as they were. Change everything without changing anything. On the other hand there might be changes on the XCCDF format level (like different order of xml elements) that are not a problem and kinda expected (unless they change the behaviour of the benchmark in any way).

#### Review Hints:

So, in addition to the changes in the build system (rather trivial), there are 2 new scripts. First and foremost, **don't even for a second think about reviewing the code itself** or its style: these scripts are one-off an will be removed in a consequent PR. They are provided for the reviewer to experiment with the conversion.

The changes are there, you don't have to convert anything.
To get the PR changes into final form you have to conduct the conversion (the order is **very** important):
```sh
./utils/convert_prodtype_into_default_profile_entry.py
./utils/convert_prodtype_remove_prodtype.sh
```
That's it. After this step the changeset is ready to be reviewed. You can build/test do whatever with the content.

Hint: to revert the changes and start over you can use git reset.

Hint: unless you clean prodtypes the `./utils/convert_prodtype_into_default_profile_entry.py` script is _re-entrant_.

If you want to narrow the scope of changes in case you'd like to experiment with a particular product you can provide `-p product_id` argument to the conversion script.

As the ultimate goal is to keep the Benchmark unchanged on the logical level you must somehow make sure of that. The special `convert_prodtype_ds_check.py` script when executed will analyze all `ssg-*-ds.xml` files and dump `ssg-*-ds.prof-stat` files in the same build directory. One way to use these files is to generate them for *new* data streams then for a reference copy of CaC from master and compare: files should be **identical**. If they are not — it is a bug.

### CODE REVIEW

**Comment by jan-cerny** [comment]
Is the reason for introduction of the concept of the "hidden profile" to not remove the rules that are already present in the data streams of the individual product and keep the backwards compatibility?

**Comment by evgenyz** [comment]
> Is the reason for introduction of the concept of the "hidden profile" to not remove the rules that are already present in the data streams of the individual product and keep the backwards compatibility?

Yes, that's the idea. Please wait before I fill the Review Hints to proceed with reviewing. It is a multi-step change and I don't want to have a PR with 2K lines floating around.

**Comment by jan-cerny** [comment]
I compared the built RHEL 9 content from current master branch as of 2023-12-14 as of HEAD 0ed28dac6b7901e798fc0ee9c8434ed5ce221593 with the built content from this PR as of HEAD a7820c6ab4b28eee7cbe7b4afd9b2e8bfa658be7.

I list using xpath the rules in the built SCAP source data stream `ssg-rhel9-ds.xml` and the set of the rules is the same in both old an new version. That's a great achievement!

However, I noticed that the OVAL is suspiciously larger. I have investigated this and I have found that the `ssg-rhel9-oval.xml` and in turn the data stream as well the new version contains many definitions which aren't present in the old definition. For example, in the new version there is OVAL definition `oval:ssg-apt_conf_disallow_unauthenticated:def:1` and this one belongs to rule `apt_conf_disallow_unauthenticated` which doesn't have `rhel9` in prodtype and isn't part of any `rhel9` profile, And the rule `apt_conf_disallow_unauthenticated` isn't present in the data stream. IOW some OVAL definitions that belong to rules that aren't shipped into the data stream unfortunately appeared.

**Comment by evgenyz** [comment]
> IOW some OVAL definitions that belong to rules that aren't shipped into the data stream unfortunately appeared.

Err, come to think about it, the way OVAL goes into DS is controlled with the same prodtype instead of cherry-picking based on actual rules that are present in the benchmark. Isn't by the way the OVAL Object Model going to fix that somehow?

File: ssg/oval_object_model/oval_container.py
Comment (evgenyz) [inline ssg/oval_object_model/oval_container.py:74]: Yep, it is a bug.

File: ssg/oval_object_model/oval_container.py
Comment (evgenyz) [inline ssg/oval_object_model/oval_container.py:229]: I know, I know. Just for sake of testing for now.

**Comment by evgenyz** [comment]
New version should be better at handling junk OVALs, but not quite there yet: tests, objects and states are still there.

**Comment by jan-cerny** [comment]
@evgenyz Regarding OVALs:

In the current master branch, the script `combine_ovals.py` loads the OVALs from files into the object model.

This script iterates over all rules in the given benchmark which most often means it iterates over the `/linux_os/guide` directory tree. It loads all OVAL files there. But, not all of them are loaded into the OVAL object model. Before loading them into object model, the script decides whether the OVAL belongs into the given product or not. If the OVAL shouldn't be there, it's not loaded into the object model and the definition isn't put in the oval-unlinked.xml.

Currently, the decision is a responsibility of the `Definition.is_applicable_for_product` method on the definition level, or on a OVAL file level the `OVALShorthand.validate` method. This is based on the contents of the `affected` elements. These elements are present in the OVAL definitions. Most often, they are populated by the `oval_metadata` Jinja macro. The jinja macros are resolved during loading of the OVAL files by `combine_ovals.py` and substituted into. This depends on the `prodtype` loaded from rules.

As you remove the prodtype from the  rules, the `affected` elements are generated differently and they stop influence the `Definition.is_applicable_for_product` return value.

Therefore I think that you need to simplify and streamline the aforementioned process.

**Comment by jan-cerny** [comment]
I think it would be nice if the `combine_ovals.py` iterated only over the rules that will be included in the final data stream instead of iterating over all rules in the benchmark directory. That'd feel simple and natural and it would reduce the complexity around `affected` elements that we experience now. However, at this moment, it isn't possible to do so, because the decision about which rules will be included in the product is done by `build_xccdf.py` and that's a later build step than `combine_ovals.py`, it even directly depends on the `oval-unlinked.xml` which is output of `combine_ovals.py`. That would have to be changed first, ideally by moving the decision moment to the resolution time, ie. `compile_all.py`. If it could be moved there, then the `combine_ovals.py` could be remade to iterate over resolved rules instead of unresolved rules.

File: ssg/oval_object_model/oval_container.py
Comment (jan-cerny) [reply ssg/oval_object_model/oval_container.py:74]: ??

Comment (evgenyz) [reply ssg/oval_object_model/oval_container.py:74]: In the original form the function `_remove_keys_from_dict` can't be used in `_call_function_for_every_component` unlike its counterpart.

**Comment by evgenyz** [comment]
Junk OVALs are done with. What is still there: extra platforms, extra variables (but these are present indiscriminately in the old-style DSes as well, if I understand it correctly).

**Comment by evgenyz** [comment]
> I think it would be nice if the `combine_ovals.py` iterated only over the rules that will be included in the final data stream instead of iterating over all rules in the benchmark directory. That'd feel simple and natural and it would reduce the complexity around `affected` elements that we experience now. However, at this moment, it isn't possible to do so, because the decision about which rules will be included in the product is done by `build_xccdf.py` and that's a later build step than `combine_ovals.py`, it even directly depends on the `oval-unlinked.xml` which is output of `combine_ovals.py`. That would have to be changed first, ideally by moving the decision moment to the resolution time, ie. `compile_all.py`. If it could be moved there, then the `combine_ovals.py` could be remade to iterate over resolved rules instead of unresolved rules.

The goal of this PR is to get rid of `prodtype` not to turn the build system inside out. So far the compiled DS is on par with the old behaviour, and I'd rather not expand the scope as it already is (will be, after adding all prodtype changes) a pretty complex change-set.

Any fine-tuning that was blocked by `prodtype` could be implemented after we get rid of the thing.

The only part I want to fix in this PR is the XCCDF variables: we just slurp them all, not caring if they are used by anything.

File: ssg/oval_object_model/oval_document.py
Comment (evgenyz) [inline ssg/oval_object_model/oval_document.py:112]: I'm not sure what was the initial idea of this function but in its current state it just hard-coded to raise an exception. For me it seems reasonable to throw a warning instead. If you @jan-cerny or @Honny1 have a better idea — I'm open to suggestions.

**Comment by evgenyz** [comment]
> The only part I want to fix in this PR is the XCCDF variables: we just slurp them all, not caring if they are used by anything.

Scratch that. It got complicated pretty fast, and I'm nowhere near the end of that fix.

In my opinion extra platforms and extra variables is something we can live with for now. It'd be easier to create a set of narrow PRs that will deal with these problems later.

@jan-cerny Can you please review the current state of this PR, keeping in mind that we focus on getting rid of `prodtype` and not on fixing everything in one go?

**Comment by evgenyz** [comment]
Also, @ggbecker and @yuumasato, do you have any opinion on this?

**Comment by jan-cerny** [comment]
@evgenyz

At the current state it can't be merged, because:
- it adds rules that shouldn't be there to the generated data stream (these are rules that have prodtype set to other product than the product currently being built)
- it removes rules that should be there from the generated data stream (these are rules that have prodtype set to the product currently being built)
- the new platform reference a non-existing component  in the href attribute of the cpe-lang:check-fact-ref element, this component belongs to different products
- CI tests don't pass

I assume that the first two points are caused by the fact that the code in this PR assumes that `prodtype`s don't exist anymore in the project when the code is merged. That leads me to thinking about how to proceed with introduction of the prodtype removal. From the top of my head I can think of two solutions:
1. Do the mass removal of the prodtype in this PR. The disadvantage is that it will shadow the changes for the reviewer and make it difficult to assess the changes.
2. Introduce only build system changes in this PR and perform the mass removal in a separate PR. This PR would have to change the code in a non-destructive way. If merged now it wouldn't have any impact on the built artifact and after the mass prodtype removal it would keep working in the new situation as well. In other words it needs to work with both situations when we have prodtype and when we don't have prodtype. The disadvantage is that it's more difficult to develop and test than the option 1.

Which of these 2  options would you prefer? Or do you have a different proposal on how to process this change?

**Comment by evgenyz** [comment]
> @evgenyz
> 
> At the current state it can't be merged, because:
> 
>     * it adds rules that shouldn't be there to the generated data stream (these are rules that have prodtype set to other product than the product currently being built)
> 
>     * it removes rules that should be there from the generated data stream (these are rules that have prodtype set to the product currently being built)

Are these cases of behavior with or without modification of prodtype and profiles (by scripts)?

>     * the new platform reference a non-existing component  in the href attribute of the cpe-lang:check-fact-ref element, this component belongs to different products

That could be a problem. Can you be a bit more specific?

>     * CI tests don't pass

That's expected until prodtype is set to implicit all.

> Which of these 2 options would you prefer? Or do you have a different proposal on how to process this change?

I'd prefer to add mass-removal to this PR (and mass introduction of default profiles). But before that I'd like us to settle on the change in the build system and check that everything is in order (i.e. change might be not ideal, but it should not be disruptive).

File: ssg/oval_object_model/oval_document.py
Comment (Honny1) [reply ssg/oval_object_model/oval_document.py:109]: This exception is that there should be no references to an OVAL component that does not appear in the OVAL document. For example, a missing state or test object.  

The same warning is issued in the [`validate_references`](https://github.com/ComplianceAsCode/content/blob/2bd80b67657e487332bce5ac70c1fbe23e891661/ssg/oval_object_model/oval_document.py#L134) function. Other cases where this expectation may be raised are when references are collected in the OVAL Definition. It should be up to the programmer to handle this case after calling the [`get_all_references_of_definition`](https://github.com/ComplianceAsCode/content/blob/2bd80b67657e487332bce5ac70c1fbe23e891661/ssg/oval_object_model/oval_container.py#L213) function.

I recommend that you keep `raise MissingOVALComponent(component_id)`.

Comment (evgenyz) [reply ssg/oval_object_model/oval_document.py:109]: The problem with the exception is that it stands in the way when the Reference object is used to actually determine what is missing and what is orphaned.

Comment (evgenyz) [reply ssg/oval_object_model/oval_document.py:109]: E.g. `_ensure_by_xccdf_referenced_oval_no_extra_def_in_oval_file` from this PR.

**Comment by jan-cerny** [comment]
> > @evgenyz
> > At the current state it can't be merged, because:
> > ```
> > * it adds rules that shouldn't be there to the generated data stream (these are rules that have prodtype set to other product than the product currently being built)
> > 
> > * it removes rules that should be there from the generated data stream (these are rules that have prodtype set to the product currently being built)
> > ```
> 
> Are these cases of behavior with or without modification of prodtype and profiles (by scripts)?

I tried it without modification of prodtype and profiles by scripts. As I wrote later in my previous comment, I assume the reason is that the code assumes that this modification has been performed before executing the build_product.

> 
> > ```
> > * the new platform reference a non-existing component  in the href attribute of the cpe-lang:check-fact-ref element, this component belongs to different products
> > ```
> 
> That could be a problem. Can you be a bit more specific?

Yes, I can. I have built the RHEL9 content and I have seen suspicious elements within the `cpe-lang:platform-specification` element. They're referencing the UOS20 CPE OVAL. For example:

```
<cpe-lang:check-fact-ref system="http://oval.mitre.org/XMLSchema/oval-definitions-5" href="ssg-uos20-cpe-oval.xml" id-ref="oval:ssg-installed_env_is_a_machine:def:1"/>
<cpe-lang:check-fact-ref system="http://oval.mitre.org/XMLSchema/oval-definitions-5" href="ssg-uos20-cpe-oval.xml" id-ref="oval:ssg-package_avahi:def:1"/>

```
> 
> > ```
> > * CI tests don't pass
> > ```
> 
> That's expected until prodtype is set to implicit all.
> 
> > Which of these 2 options would you prefer? Or do you have a different proposal on how to process this change?
> 
> I'd prefer to add mass-removal to this PR (and mass introduction of default profiles). But before that I'd like us to settle on the change in the build system and check that everything is in order (i.e. change might be not ideal, but it should not be disruptive).

OK. So let me check how it works with the prodtypes removed by the one-off scripts.

**Comment by jan-cerny** [comment]
I have checked the output of the build system with the prodtypes removed by the one-off scripts. The main concern is the aforementioned problem with the `<cpe-lang:check-fact-ref>` elements which also occurs in this situation. I think that needs to be adressed. Other problems are stray platforms and stray OCIL checks but we can survive these.

File: ssg/oval_object_model/oval_document.py
Comment (Honny1) [reply ssg/oval_object_model/oval_document.py:109]: To get a list of missing references. I would suggest creating a method of OVAL Document that provides the missing references and bypasses the exception invocation.

File: ssg/build_renumber.py
Comment (Honny1) [inline ssg/build_renumber.py:272]: If I understand the function correctly. You only want to preserve the OVAL definitions referenced from XCCDF in the OVAL document. I would suggest generating all references for OVAL definitions referenced from XCCDF (aka `references_from_xccdf_to_keep`) and then passing those to the `self.oval_document.keep_referenced_components(references_from_xccdf_to_keep)` method. Then only the OVAL definitions referenced in the XCCDF should be present in the OVAL document, along with all their components. 

```suggestion
    def _ensure_by_xccdf_referenced_oval_no_extra_def_in_oval_file(self):
        # Remove all OVAL checks that are not referenced by XCCDF Rules (checks)
        # or internally via extend-definition

        xccdf_oval_check_refs = [name for name in self._get_list_of_names_of_oval_checks()]
        document_def_keys = list(self.oval_document.definitions.keys())

        references_from_xccdf_to_keep = OVALDefinitionReference()
        for def_id in document_def_keys:
            if def_id in xccdf_oval_check_refs:
                oval_def_refs = self.oval_document.get_all_references_of_definition(def_id)
                references_from_xccdf_to_keep += oval_def_refs

        self.oval_document.keep_referenced_components(references_from_xccdf_to_keep)
```

File: ssg/oval_object_model/oval_document.py
Comment (evgenyz) [reply ssg/oval_object_model/oval_document.py:109]: Okay, I've played with this a bit more, and I think we could be able to just live with the Exeption after fixing the last part that gives us this problem (https://github.com/ComplianceAsCode/content/pull/11432). This, after all, is good mechanism that prevents content authors from doing nasty things (see the PR).

File: ssg/build_renumber.py
Comment (evgenyz) [reply ssg/build_renumber.py:272]: Noice! I'll check it out, thanks.

File: shared/applicability/oval/installed_app_is_eks.xml
Comment (yuumasato) [inline shared/applicability/oval/installed_app_is_eks.xml:8]: Why was `multi_platform_ocp` added here?
Is this required for the content to build?

File: shared/applicability/oval/installed_app_is_eks_node.xml
Comment (yuumasato) [inline shared/applicability/oval/installed_app_is_eks_node.xml:7]: Ditto

**Review by yuumasato — COMMENTED** [review]
This is not a complete review...but these caught my eyes right now.

File: shared/applicability/oval/installed_app_is_eks.xml
Comment (evgenyz) [reply shared/applicability/oval/installed_app_is_eks.xml:8]: Yes. The longer answer is: we technically need both these checks (it is a CPE check) in the DS to tell platforms apart.

Comment (evgenyz) [reply shared/applicability/oval/installed_app_is_eks.xml:8]: It is symmetrical to this problem: https://github.com/ComplianceAsCode/content/pull/10610

File: shared/applicability/oval/installed_app_is_eks_node.xml
Comment (evgenyz) [reply shared/applicability/oval/installed_app_is_eks_node.xml:7]: Well, ditto :)

File: shared/applicability/oval/installed_app_is_eks.xml
Comment (evgenyz) [reply shared/applicability/oval/installed_app_is_eks.xml:8]: Better yet: https://github.com/ComplianceAsCode/content/pull/11436

**Comment by evgenyz** [comment]
> I have checked the output of the build system with the prodtypes removed by the one-off scripts. The main concern is the aforementioned problem with the `<cpe-lang:check-fact-ref>` elements which also occurs in this situation. I think that needs to be adressed.

Fixed.

File: ssg/build_renumber.py
Comment (evgenyz) [reply ssg/build_renumber.py:272]: Wonderful. I've simplified the change-set.

File: ssg/entities/profile_base.py
Comment (jan-cerny) [inline ssg/entities/profile_base.py:117]: I think it shouldn't return a comment.

File: shared/applicability/oval/installed_app_is_eks.xml
Comment (yuumasato) [reply shared/applicability/oval/installed_app_is_eks.xml:8]: I see, thanks.

File: ssg/entities/profile_base.py
Comment (evgenyz) [reply ssg/entities/profile_base.py]: So, here is the problem. It is expected that this function will return an `ET.Element` (the comment is actually `ET.Element(Comment)`).

This comment will be present in the `ssg-blah-xccdf.xml` file, but would be filtered out in the final `ssg-blah-ds.xml` data stream.

That still is the best idea I have about how to handle it.

**Comment by jan-cerny** [comment]
@evgenyz I have found another problem. The rules in the built data stream started to be applicable on other platforms than they should be. For example, I have built the SLES 15 product , run oscap and the rules are evaluated as applicable on my Fedora machine.

**Comment by evgenyz** [comment]
> @evgenyz I have found another problem. The rules in the built data stream started to be applicable on other platforms than they should be. For example, I have built the SLES 15 product , run oscap and the rules are evaluated as applicable on my Fedora machine.

Eagle eye! Fixed.

**Comment by yuumasato** [comment]
The `ocp4` and `rhcos4` contents seem ok and are working correctly.

**Comment by jan-cerny** [comment]
The RHEL9 and SLE15 content looks fine. Also I no longer experience the aforementioned applicability problem.

I think you can proceed with a mass removal. Let's see how CI thinks.

**Comment by evgenyz** [comment]
The `test-no-new-global-applicable-rules` test expectantly failing.

**Comment by evgenyz** [comment]
@ComplianceAsCode/oracle-maintainers @ComplianceAsCode/red-hatters @ComplianceAsCode/suse-maintainers @ComplianceAsCode/ubuntu-maintainers Please give me an ACK that you understand what will happen when this PR is merged. Or — speak up: voice your concerns, ask questions etc. We can even have a dedicated online meeting on the topic of this change if needed.

**Comment by evgenyz** [comment]
@jan-cerny I've managed to fix all problems related to this PR, and the fixes are simple enough to be included right here, IMHO. But if you wish I can extract them as a separate PR.

I believe we are good to go with the merge.

**Review by dodys — APPROVED** [review]
lgtm
since this is a big change, are we doing a major version upgrade?

**Comment by evgenyz** [comment]
> lgtm since this is a big change, are we doing a major version upgrade?

We are not changing anything user-facing. Don't think we need a version bump.

**Review by Mab879 — APPROVED** [review]
I think these changes will be good.

Thanks for your hard work.

**Comment by Mab879** [comment]
LGTM.

I am overriding the CODEOWERS as acknowledgement of this change is given [here](https://github.com/ComplianceAsCode/content/issues/11382#issuecomment-1857665255).

I am overriding the Automatus tests as they trying to run everything since almost every rule was touched.

**Comment by Mab879** [comment]
Also waving code climate as these issues are in the temporary scripts.

**Pull request merged**
